lib/api/CDataFrameAnalysisRunner.cc

```diff
@@ -36,6 +37,19 @@ CDataFrameAnalysisRunner::~CDataFrameAnalysisRunner() {
     this->waitToFinish();
 }
 
+SMemoryUsageEstimationResult CDataFrameAnalysisRunner::estimateMemoryUsage() const {
+    std::size_t numberRows{m_Spec.numberRows()};
+    std::size_t numberColumns{m_Spec.numberColumns() + this->numberExtraColumns()};
+    std::size_t maximumNumberPartitions{
+        static_cast<std::size_t>(std::sqrt(static_cast<double>(numberRows)) + 0.5)};
+
+    std::size_t memoryUsageWithOnePartition{
+        this->estimateMemoryUsage(numberRows, numberRows, numberColumns)};
+    std::size_t memoryUsageWithMaxPartitions{
+        this->estimateMemoryUsage(numberRows, numberRows / maximumNumberPartitions, numberColumns)};
```

Comment:
Whilst I think it is very likely that the memory usage decreases monotonically with increasing number of partitions I wonder whether we should just factor out the loop from `CDataFrameAnalysisRunner::computeAndSaveExecutionStrategy` into a shared private function, say `estimateMinimumMemoryUsage`, and return the minimum memory limit calculated in the loop. This means that these results keeps step with the logic which is applied there, which it must. Also, the compute cost of the loop should be very low compared to overheads. A related question mark is whether this should respect `m_Spec.diskUsageAllowed()`.

Also this should apply rounding I think, as per [this](https://github.com/elastic/ml-cpp/blob/master/lib/api/CDataFrameAnalysisRunner.cc#L83) line.

Reply:
Note we discussed this further offline and decided that we don't want to respect `m_Spec.diskUsageAllowed()` here in which case there is less reason to try and share this code with `CDataFrameAnalysisRunner::computeAndSaveExecutionStrategy`.

Reply:
IMO it is odd to output the estimate to the nearest byte (imagine reporting you need 49991231 bytes to run this analysis). I think it also gives the wrong impression about how accurate the memory is. So, I'd be inclined to round up to the nearest MB/100.

Reply:
When `model_memory_limit` is configured via the Elasticsearch `ByteSizeValue` functionality it must be specified as a whole number of some unit.  So something like `"4.56mb"` will not parse using the functionality of `ByteSizeValue`.  Therefore I think we should pick a unit and round up to whole multiples of that unit.  Otherwise we're just creating a need to do extra processing further along the chain.

Anomaly detection jobs have to have `model_memory_limit`s that are whole numbers of megabytes.  We could stick with that for data frame analytics or go to kilobytes I guess.  I am not sure there is _that_ much benefit in going to kilobytes because at such a small level the overhead of running the process dominates the memory required by the analytics.  But if people want something more granular than megabytes then I'm happy with kilobytes.

Reply:
Done (rounding up to the nearest kilobytes).
Is it ok to still return the results as just bytes? On Java side I can just create ByteSizeValue using public constructor:
    public ByteSizeValue(long bytes) {...}

Reply:
Scratch that, I'll just use strings in both Java-cpp protocol and public API

Reply:
Done.